jib-core/src/main/java/com/google/cloud/tools/jib/http/Connection.java

```diff
@@ -114,87 +109,157 @@ private static void addProxyCredentials(ApacheHttpTransport transport, String pr
+      Consumer<LogEvent> logger) {
+    this(
+        enableHttpAndInsecureFailover,
+        sendAuthorizationOverHttp,
+        logger,
+        Connection::getSecureHttpTransport,
+        Connection::getInsecureHttpTransport);
   }
 
-  @Override
-  public void close() throws IOException {
-    if (httpResponse == null) {
-      return;
-    }
-
-    httpResponse.disconnect();
+  @VisibleForTesting
+  Connection(
+      boolean enableHttpAndInsecureFailover,
+      boolean sendAuthorizationOverHttp,
+      Consumer<LogEvent> logger,
+      Supplier<HttpTransport> secureHttpTransportFactory,
+      Supplier<HttpTransport> insecureHttpTransportFactory) {
+    this.enableHttpAndInsecureFailover = enableHttpAndInsecureFailover;
+    this.sendAuthorizationOverHttp = sendAuthorizationOverHttp;
+    this.logger = logger;
+    this.secureHttpTransportFactory = secureHttpTransportFactory;
+    this.insecureHttpTransportFactory = insecureHttpTransportFactory;
   }
 
   /**
    * Sends the request with method GET.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response get(Request request) throws IOException {
-    return send(HttpMethods.GET, request);
+  public Response get(URL url, Request request) throws IOException {
+    return call(HttpMethods.GET, url, request);
   }
 
   /**
    * Sends the request with method POST.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response post(Request request) throws IOException {
-    return send(HttpMethods.POST, request);
+  public Response post(URL url, Request request) throws IOException {
+    return call(HttpMethods.POST, url, request);
   }
 
   /**
    * Sends the request with method PUT.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response put(Request request) throws IOException {
-    return send(HttpMethods.PUT, request);
+  public Response put(URL url, Request request) throws IOException {
+    return call(HttpMethods.PUT, url, request);
   }
 
   /**
    * Sends the request.
    *
    * @param httpMethod the HTTP request method
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if building the HTTP request fails.
    */
-  public Response send(String httpMethod, Request request) throws IOException {
-    Preconditions.checkState(httpResponse == null, "Connection can send only one request");
+  public Response call(String httpMethod, URL url, Request request) throws IOException {
+    if (!isHttpsProtocol(url) && !enableHttpAndInsecureFailover) {
```

Comment:
I think it's worth explicitly calling out that we're following the approach taken by Docker and always contacting HTTPS and then retrying as HTTP (if allowed).

Could we make the failure-over check be a method that takes the URL (`canFailOverToHttp(url)`?) so that we can swap out a different implementation (e.g., a table lookup populated from `.docker/daemon.json`) without changing this code?

Reply:
sgtm